### PR TITLE
Add Ddx and Ddy builtins, and implement for all backends

### PR DESCRIPTION
**Pull request by tgjones**
Coming from an HLSL background, I tend to favour the HLSL names for these things, but obviously let me know if you'd prefer something else :)

### CODE REVIEW

**Comment by mellinoe** [comment]
Again it looks good overall, and I prefer the HLSL syntax as well, generally speaking.

Ddx and Ddy can only be used in the fragment stage, so we need to catch that and fail early. I have a patch that fixes it: https://github.com/mellinoe/ShaderGen/commit/d5fda2a9bf57c7955ff4b5991a24da88c5a69413

I haven't tested the HLSL compilation (working on my Mac tonight), but it looks correct.

**Comment by tgjones** [comment]
I've cherry-picked d5fda2a - thank you.

File: src/ShaderGen/ShaderMethodVisitor.cs [hunk 1]
```diff
@@ -205,6 +205,12 @@ public override string VisitInvocationExpression(InvocationExpressionSyntax node
                 SymbolInfo symbolInfo = GetModel(node).GetSymbolInfo(ins);
                 string type = symbolInfo.Symbol.ContainingType.ToDisplayString();
                 string method = symbolInfo.Symbol.Name;
+
+                if (type == "ShaderGen.ShaderBuiltins")
```
Comment (tzachshabtay) [inline src/ShaderGen/ShaderMethodVisitor.cs]: Can't `nameof` be used here?

Comment (mellinoe) [reply src/ShaderGen/ShaderMethodVisitor.cs]: `nameof` probably could / should be used in a lot more places, but it's kind of hit-or-miss in the current code. I'm fine with leaving it like this for now, although we should do a cleanup sweep sometime.

**Pull request merged**